src/CpuAffinity.cc

```diff
@@ -64,3 +42,38 @@ CpuAffinityCheck()
     }
 }
 
+/// Makes CPU affinity of this process match configured CPU affinity.
+/// Assumes that we have never set the CPU affinity or have cleared it.
+static void
+CpuAffinityInit()
+{
+    if (IamPrimaryProcess())
+        CpuAffinityCheck();
+    Must(!TheCpuAffinitySet);
+    if (Config.cpuAffinityMap) {
+        const int processNumber = InDaemonMode() ? KidIdentifier : 1;
+        TheCpuAffinitySet = Config.cpuAffinityMap->calculateSet(processNumber);
+        if (TheCpuAffinitySet)
+            TheCpuAffinitySet->apply();
+    }
+}
+
+class CpuAffinityRr : public RegisteredRunner
+{
+public:
+    virtual void useConfig() override {
+        CpuAffinityInit();
+    }
+
+    virtual void syncConfig() override {
+        if (TheCpuAffinitySet) {
+            TheCpuAffinitySet->undo();
+            delete TheCpuAffinitySet;
+            TheCpuAffinitySet = nullptr;
+        }
+        CpuAffinityInit();
```

Comment:
Please do not clear the current affinity _before_ checking whether the new affinity is valid. Eventually, CpuAffinityCheck() will reject invalid configurations. When we reject an invalid configuration during reconfiguration, the old set affinity should stay intact.

It is hard to see what is going on in the diff because you moved the essentially unchanged code, but the old code did not have that flaw. It is easy to fix though AFAICT: Just move the `if` statement calling CpuAffinityCheck() into the beginning of useConfig() and syncConfig() methods. Eventually, there will be a checkConfig() event, and that duplicated code will go into that single event handler.